core/arch/arm/plat-hikey/main.c

```diff
@@ -93,3 +108,108 @@ void console_flush(void)
 {
 	pl011_flush(console_base());
 }
+
+vaddr_t peripherals_base(paddr_t pa)
+{
+	void *va;
+
+	if (cpu_mmu_enabled()) {
+		va = phys_to_virt(pa, MEM_AREA_IO_NSEC);
+		return (vaddr_t)va;
+	}
+	return (vaddr_t)pa;
+}
+
+#ifdef CFG_SPI
+void spi_init(void)
+{
+	uint32_t shifted_val, read_val;
+	vaddr_t peri_base = peripherals_base(PERI_BASE);
+	vaddr_t pmx0_base = peripherals_base(PMX0_BASE);
+	vaddr_t pmx1_base = peripherals_base(PMX1_BASE);
+
+	DMSG("take SPI0 out of reset\n");
+	shifted_val = PERI_RST3_SSP;
+	/*
+	 * no need to read PERI_SC_PERIPH_RSTDIS3 first
+	 * as all the bits are processed and cleared after writing
+	 */
+	write32(shifted_val, peri_base + PERI_SC_PERIPH_RSTDIS3);
+	DMSG("PERI_SC_PERIPH_RSTDIS3: 0x%x\n", read32(peri_base
+		+ PERI_SC_PERIPH_RSTDIS3));
+
+	/*
+	 * wait until the requested device is out of reset
+	 * and ready to be used
+	 */
+	do {
+		read_val = read32(peri_base + PERI_SC_PERIPH_RSTSTAT3);
+	} while (read_val & shifted_val);
+	DMSG("PERI_SC_PERIPH_RSTSTAT3: 0x%x\n", read_val);
+
+	DMSG("enable SPI clock\n");
+	/*
+	 * no need to read PERI_SC_PERIPH_CLKEN3 first
+	 * as all the bits are processed and cleared after writing
+	 */
+	shifted_val = PERI_CLK3_SSP;
+	write32(shifted_val, peri_base + PERI_SC_PERIPH_CLKEN3);
+	DMSG("PERI_SC_PERIPH_CLKEN3: 0x%x\n", read32(peri_base
+		+ PERI_SC_PERIPH_CLKEN3));
+
+	DMSG("PERI_SC_PERIPH_CLKSTAT3: 0x%x\n", read32(peri_base
+		+ PERI_SC_PERIPH_CLKSTAT3));
+
+	/*
+	 * gpio6_2 can be configured as pinmux_spi, in which case the hw
+	 * will control the chip select pin and we don't have to manually
+	 * do it, but hw will pulse it between each data word transfer,
+	 * which will not work with all clients. There seems to be no
+	 * option to configure it to stay enabled for the total duration
+	 * of the transfer.
+	 * ref: infocenter.arm.com/help/topic/com.arm.doc.ddi0194h/CJACFAFG.html
+	 */
+	DMSG("configure gpio6_{0,1,3} as SPI\n");
+	DMSG("configure gpio6_2 as GPIO\n");
+	write32(PINMUX_SPI, pmx0_base + PMX0_IOMG104); /* 0xF70101A0 */
+	write32(PINMUX_SPI, pmx0_base + PMX0_IOMG105); /* 0xF70101A4 */
+	write32(PINMUX_GPIO, pmx0_base + PMX0_IOMG106); /* 0xF70101A8 */
+	write32(PINMUX_SPI, pmx0_base + PMX0_IOMG107); /* 0xF70101AC */
+
+	DMSG("configure gpio6_{0:3} as nopull\n");
+	write32(PINCFG_NOPULL, pmx1_base + PMX1_IOCG104); /* 0xF70109B0 */
+	write32(PINCFG_NOPULL, pmx1_base + PMX1_IOCG105); /* 0xF70109B4 */
+	write32(PINCFG_NOPULL, pmx1_base + PMX1_IOCG106); /* 0xF70109B8 */
+	write32(PINCFG_NOPULL, pmx1_base + PMX1_IOCG107); /* 0xF70109BC */
+
+#ifdef CFG_SPI_TEST
+	spi_test();
+#endif
+}
+#endif
+
+static TEE_Result peripherals_init(void)
+{
+	uint8_t data;
+	vaddr_t pmussi_base = peripherals_base(PMUSSI_BASE);
+	vaddr_t enable_ldo17_22 = pmussi_base + (0x02f << 2);
```

Comment:
`0x2f` looks like a magic number. Please fix the other magic numbers at the same time.